### PR TITLE
Add SAVE_CURRENT_POSITION, RESTORE_POSITION

### PR DESCRIPTION
**Pull request by wizhippo**
Add SAVE_CURRENT_POSITION, RESTORE_POSITION commands

This allows you to save position and return to it later. Useful for when
pausing SD print.

ex:

M25
SAVE_CURRENT_POSITION NAME=sdpause
G91
M83
G1 Z+5 E-3 F4500
M82
G90
G1 X0 Y0

//do something

RESTORE_POSITION NAME=sdpause X=0 Y=0 F=4500
RESTORE_POSITION NAME=sdpause Z=0 E=0 F=0
M24

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks.  However, I'd rather not introduce rare and exotic g-code commands.  I have a strong preference for commands like "SAVE_CURRENT_POSITION NAME=xyz" and "RESTORE_POSITION NAME=xyz".  Also, there's no need to modify gcode.py - instead, create a "plugin" by adding a new file to klippy/extras/ - see the klippy/extras/probe.py file as an example.

**Comment by wizhippo** [comment]
Better?

**Comment by wizhippo** [comment]
I'm don't want to repeat the code G1/G0 already perform but if that is how you prefer it I will.

Also how can I create a command that it not a GCODE command that behave like one such as have the params passed to it like X0 Y0 parsed the same as the cmd_Gxx commands?

**Comment by wizhippo** [comment]
Found what I wanted to do in octoprint.

**Comment by wizhippo** [comment]
Octoprint messes up with extrusion even though it has been set to relative so I am reopening this.

This works as is except if a relative move is made as in the example in the description after a save, on restore the wrong position is used.

File: klippy/extras/position_recall.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                                    desc=self.cmd_RESTORE_POSITION_help)
+    cmd_SAVE_CURRENT_POSITION_help = "Save the current position to a slot"
+    def cmd_SAVE_CURRENT_POSITION(self, params):
+        name = self.gcode.get_str('NAME', params, 'default')
+        self.slots[name] = (self.toolhead.get_position(), self.gcode.speed)
+    cmd_RESTORE_POSITION_help = "Move to the position previously saved in a slot"
+    def cmd_RESTORE_POSITION(self, params):
+        name = self.gcode.get_str('NAME', params, 'default')
+        if name not in self.slots:
+            raise error("Invalid slot '%s' not previously set" % (name,))
```
Comment (KevinOConnor) [inline klippy/extras/position_recall.py]: Unless I'm missing something, that should be "raise self.gcode.error(...)".

File: klippy/extras/position_recall.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        pos = self.toolhead.get_position()
+        changed = False
+        for axis in params.keys():
+            if axis in self.gcode.axis2pos:
+                pos[self.gcode.axis2pos[axis]] = self.slots[name][0][self.gcode.axis2pos[axis]] + self.gcode.get_float(axis, params, 0.)
```
Comment (KevinOConnor) [inline klippy/extras/position_recall.py]: axis2pos is an internal field (purely for optimization) of the gcode class.  Consider using "for axis_pos, axis in enumerate("XYZ"):"

The member variables of a class can change and it's a pain to try and track down all external users.  I'd prefer if only exported methods and no members (ie, axis2pos, speed, speed_factor) were used.  It's fine to add additional fields to gcode.get_status() or to add an additional access method.

File: klippy/extras/position_recall.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                changed = True
+        if not changed:
+            pos = self.slots[name][0]
+        if 'F' in params.keys():
```
Comment (KevinOConnor) [inline klippy/extras/position_recall.py]: Minor, but I think SPEED=10 would be preferable to F=600.

**Comment by KevinOConnor** [comment]
As a high-level question - is this for scripting use, or is this something you were thinking people would manually type in?  If for scripting use, will it be necessary to also backup and restore the G90/G91 and G92 state?

File: klippy/extras/position_recall.py (hunk 1 above)
Comment (wizhippo) [reply klippy/extras/position_recall.py]: Error is not part of the class and defined at the top level of the module and not part of the PositionRecall class .  Followed your example in another file.  Can change if you would like.

**Comment by wizhippo** [comment]
I intend to use this with octoprint's pause and resume gcode snippets.  I want when I pause the head to move up and goto x0y0.

Then when I resume I want to purge then move xy first to the stored position, then move z and e back and resume where left off.  If after a save any G90/G91 and G92 etc have been called, I would expect the user to have to restore the those changes prior to a restore though E could be troublesome.

I tried the example octoprint http://docs.octoprint.org/en/master/features/gcode_scripts.html#more-nifty-pause-and-resume but it does not work as expected.  It messes up on restoring the last extrusion position.  Having the ability to save and restore position would make this work as expected without having to rely on octoprint getting the correct positions prior to running the script.  At least that is the thought.

Still have to figure out best way to handle restoring the extrude position in case between calls the user primed etc.

Perhaps looking at the example script you may spot why the extrude does not restore to the correct last position and this PR would not be needed.

**Comment by KevinOConnor** [comment]
On Fri, Mar 09, 2018 at 08:52:50AM -0800, Douglas Hammond wrote:
> I intend to use this with octoprint's pause and resume gcode snippets.  I want when I pause the head to move up and goto x0y0.
> 
> Then when I resume I want to purge then move xy first to the stored position, then move z and e back and resume where left off.  If after a save any G90/G91 and G92 etc have been called, I would expect the user to have to restore the those changes prior to a restore though E could be troublesome.

If the gcode file being printed is always using absolute XYZ
coordinates and absolute E coordinates then you just need to make sure
to set absolute coordinates during the resume.  However, if the print
is in relative mode then setting absolute mode will mess it up.
Another exmample of how g-code is incredibly painful to use.

> I tried the example octoprint http://docs.octoprint.org/en/master/features/gcode_scripts.html#more-nifty-pause-and-resume but it does not work as expected.  It messes up on restoring the last extrusion position.  Having the ability to save and restore position would make this work as expected without having to rely on octoprint getting the correct positions prior to running the script.  At least that is the thought.

Okay.  If you want to do a full save/restore of all state, then
probably doing it in gcode.py is the best way.  (Contrary to my
previous statements.)

> Still have to figure out best way to handle restoring the extrude position in case between calls the user primed etc.
> 
> Perhaps looking at the example script you may spot why the extrude does not restore to the correct last position and this PR would not be needed.

I ran a quick test and it looks like OctoPrint is correctly parsing
the values.

I'm not really sure what you're trying to do though.  You don't want
to actually move the extruder on restore (if the user primed the
extruder by pushing out 10mm of filament, you don't want to retract
the extruder by 10mm before resuming).  I guess what you want is to
reset the g-code absolute coordinates so that new moves are relative
to the new (post primed) position?  If so, I think the OctoPrint
example looks correct.  What did it do that was incorrect?

-Kevin

**Comment by wizhippo** [comment]
Octoprint example tried to extrude 300+mm on resume.

Testing all I did was start printing a benchy.  Paused in octoprint then resumed.
As expected the head lifted and retracted accordingly. Then moved to X0 Y0. Then on resume it primed  but then tried to extrude some crazy amount and failed thanks to the over extrude length protection. 

I'll try it once more time tonight to confirm and let you know.

**Comment by KevinOConnor** [comment]
On Fri, Mar 09, 2018 at 09:52:43AM -0800, Douglas Hammond wrote:
> Octoprint example tried to extrude 300+mm on resume.
> 
> Testing all I did was start printing a benchy.  Paused in octoprint then resumed.
> As expected the head lifted and retracted accordingly. Then moved to X0 Y0. Then on resume it primed  but then tried to extrude some crazy amount and failed thanks to the over extrude length protection. 

I suspect OctoPrint wants the values returned from M114 to be relative
to the last G92 command.  If you try on the work-m114-20180309 branch
does it improve your results?

cd ~/klipper ; git fetch ; git checkout origin/work-m114-20180309 ; sudo service klipper restart

-Kevin

**Comment by wizhippo** [comment]
Justed tested on master and not sure what chagned but the example from octoprint is working

**Comment by KevinOConnor** [comment]
On Thu, Mar 15, 2018 at 03:01:31PM -0700, Douglas Hammond wrote:
> Justed tested on master and not sure what chagned but the example from octoprint is working

So, master works, but work-m114-20180309 does not?  I'm surprised as I
think the change to M114 is needed.

-Kevin

**Comment by wizhippo** [comment]
Ok did not work as I tried again later with master and got the error on exceeding the max extrusion.

Tried again on the branch you mention and I have the same issue.

**Comment by wizhippo** [comment]
[klippy.log](https://github.com/KevinOConnor/klipper/files/1817423/klippy.log)
Here is the log.

**Comment by KevinOConnor** [comment]
On Thu, Mar 15, 2018 at 04:36:07PM -0700, Douglas Hammond wrote:
> Here is the log.

Ah, I didn't think about extrude factor in the new M114 code.  I think
that should be fixed now.

cd ~/klipper ; git fetch ; git checkout origin/work-m114-20180309 ; sudo service klipper restart

-Kevin

**Comment by wizhippo** [comment]
Looks like work-m114-20180309 works now.  Thank you for working on this.

**Comment by KevinOConnor** [comment]
Thanks - I committed it to the master branch (451f7d56).  I do like your idea of introducing a new command to save and restore state/position.  I'm inclined to do that after releasing v0.6.0 though.

**Comment by wizhippo** [comment]
Have working solution in octoprint for now.  Thanks again.

**Pull request closed without merging**